### PR TITLE
making sure that n_trials will run in parallel mode

### PR DESCRIPTION
**Pull request by Ohadkaiser**
## Motivation
Suggesting solution to #3518 and #3938.
In this way you will be able to make sure that the number of trials that is delivered in the parameter n_trials will be exact when running in parallel. 

## Description of the changes
Added the ability to get the amount of trials in the study and to stop after the number of trials is reached

### CODE REVIEW

**Comment by not522** [comment]
Your suggestion seems to break the following code. (Original: https://github.com/optuna/optuna/issues/2907#issuecomment-914610505)

```python
study.optimize(objective, n_trials=100) # not good objective value yet
study.optimize(objective, n_trials=100) # perform additional 100 trials to get a better value
```

I recommend not changing the behavior of existing user codes. Instead, it is better to add a new feature.

**Comment by HideakiImamura** [comment]
@not522 Could you review this PR?

File: optuna/study/_optimize.py [hunk 1]
```diff
@@ -122,60 +229,6 @@ def _optimize(
...
-        if callbacks is not None:
-            for callback in callbacks:
-                callback(study, frozen_trial)
-
-        if progress_bar is not None:
-            progress_bar.update((datetime.datetime.now() - time_start).total_seconds())
-
-    study._storage.remove_session()
-
-
```
Comment (nzw0301) [inline optuna/study/_optimize.py:178]: Could you revert this change? It makes reviews harder.

Comment (Ohadkaiser) [reply optuna/study/_optimize.py:178]: I changed the order so that I could use `_optimize_sequential` as a default parameter for `_optimize` function

**Comment by nzw0301** [comment]
TBH, I'm not sure that introducing a new public method: `optimize_parallel` and its private method that shares many lines with the existing method for this scenario is the best approach.

**Comment by Ohadkaiser** [comment]
> TBH, I'm not sure that introducing a new public method: `optimize_parallel` and its private method that shares many lines with the existing method for this scenario is the best approach.

Do you have any suggestions that won't break the current optimize method implementation?

**Comment by nzw0301** [comment]
I might still misunderstand the situation, but for me, [`MaxTrialsCallback`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.MaxTrialsCallback.html) with `states=None` is supposed to be enough.

**Comment by Ohadkaiser** [comment]
> I might still misunderstand the situation, but for me, [`MaxTrialsCallback`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.MaxTrialsCallback.html) with `states=None` is supposed to be enough.

I'm trying to fix #3518, So it won't be enough in my opinion

**Comment by nzw0301** [comment]
I've left comments on the original issue as in https://github.com/optuna/optuna/issues/3518#issuecomment-1236119954. Could you give some examples that my suggestion cannot resolve?
